File: 229-majority-element-ii/majority-element-ii.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var majorityElement = function(nums) {
    let count1 = 0, count2 = 0, cand1 = null, cand2 = null;
    for (let num of nums) {
        if (num === cand1) count1++;
        else if (num === cand2) count2++;
        else if (count1 === 0) { cand1 = num; count1 = 1; }
        else if (count2 === 0) { cand2 = num; count2 = 1; }
        else { count1--; count2--; }
    }
    count1 = count2 = 0;
    for (let num of nums) {
        if (num === cand1) count1++;
        else if (num === cand2) count2++;
    }
    let res = [];
    if (count1 > Math.floor(nums.length / 3)) res.push(cand1);
    if (count2 > Math.floor(nums.length / 3)) res.push(cand2);
    return res;
};

File: 229-majority-element-ii/majority-element-ii.java
class Solution {
    public List<Integer> majorityElement(int[] nums) {
        List<Integer> list = new ArrayList<>();

        int v1 =0;
        int cand1=0;
        int v2= 0;
        int cand2=0;

        for(int num:nums){
            if(cand1==num) v1++;
            else if(cand2==num) v2++;
            else if(v1==0) {cand1=num;v1=1;}
            else if(v2==0) {cand2=num;v2=1;}
            else {v1--; v2--;}
        }

        int cnt1=0;
        int cnt2=0;

        for(int num:nums){
            if(num==cand1) cnt1++;
            else if(num==cand2) cnt2++;
        }
        System.out.println(cand1);
                System.out.println(cand2);
        if(cnt1>nums.length/3) list.add(cand1);
        if(cnt2>nums.length/3) list.add(cand2);

        return list;
    }
}